### ISSUE TITLE
`mutable_key_type` doesn't recursively consider `ignore_interior_mutability` config

### ISSUE THREAD

**Issue by Alexendoo** [rust-lang/rust-clippy#9801]
Adding `bytes::Bytes` doesn't catch the FP in https://github.com/rust-lang/rust-clippy/issues/5325#issuecomment-1284791664:

```rust
use http::header::HeaderName;
use std::{collections::HashMap, str::FromStr};

fn main() {
    let mut foo = HashMap::new();
    foo.insert(HeaderName::from_str("hi").unwrap(), 3);
}
```

I *think* what's happening is that the check for `self.ignore_mut_def_ids.contains` is shallow, but `ty.is_freeze` is deep. But it might be something else as the path from `HeaderName` -> `Bytes` is fairly complex: https://docs.rs/http/latest/src/http/header/name.rs.html#33-35

cc @llogiq
